lib/src/time_pattern.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright 2024 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Provides a TimeExpression type that represents a range of time.
+
+use chrono::{DateTime, TimeZone};
+use chrono_english::{parse_date_string, DateError, Dialect};
+use thiserror::Error;
+
+use crate::backend::Timestamp;
+
+/// Error occurred during time pattern parsing.
+#[derive(Debug, Error)]
+pub enum TimePatternParseError {
+    /// Unknown pattern kind is specified.
+    #[error(r#"Invalid time pattern kind "{0}:""#)]
+    InvalidKind(String),
+    /// Failed to parse input UI path.
+    #[error(transparent)]
+    ParseError(#[from] DateError),
+}
+
+/// Represents an range of dates and times that revisions may be matched
+/// against.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum TimePattern {
+    /// Represents all times at or after the given instant in time.
+    AtOrAfter(Timestamp),
+    /// Represents all times before, but not including, the given instant in
+    /// time.
+    Before(Timestamp),
+}
+
+impl TimePattern {
+    /// Parses a string into a TimePattern.
+    pub fn from_str_kind<Tz: TimeZone>(
+        s: &str,
+        kind: Option<&str>,
+        now: DateTime<Tz>,
```

Comment:
nit: maybe use `backend::Timestamp` for consistency? (You might want to move time_util from cli crate.)

If this is for future DST support, some comment will help.

Reply:
This is for DST support, so I'll add a comment. If the local offset changed between when the command is run and the date string (like "yesterday"), the full timezone would be needed to understand what "yesterday at midnight" was.